### ISSUE TITLE
Ability to execute VBA code from a VBA addin

### ISSUE THREAD

**Issue by WaynePhillipsEA** [WaynePhillipsEA/twinbasic#818]
A question came up at the Access DevCon today as to whether we can trigger / execute VBA code from within a VBIDE addin written in twinBASIC.  

For those of us involved in the Rubberduck VBA project, this probably brings back some fond memories(!), as it's not as easy as it first appears :) 

This would probably be best implemented as a package.  It should be relatively straight-forward to migrate the code over from RD to handle this and thus support every VBA host.

**Comment by Greedquest** [comment]
Way ahead of you :wink:: https://codereview.stackexchange.com/questions/274532/low-level-vba-hacking-making-private-functions-public

Well a little ahead, that's VBA not tb. I didn't realise it was your code I was translating!

**Comment by WaynePhillipsEA** [comment]
Nice!  That looks like a good starting point.  I would think we can simplify that quite a bit for the tB version ;)

**Comment by bclothier** [comment]
One potential opportunity for tB to make it even easier is to allow native method of calling `IDispatch::Invoke` directly without all the extra hacks. 

We may also want to look at exposing the type library interfaces natively in tB. I think that if we enable the interfaces natively, then everything else becomes much more easier since we can now directly manipulate the type infos, even at runtime.

**Comment by Greedquest** [comment]
Yeah declaring com interfaces with udts, pointer offsets and DispCallFunc is like 90% of the difficulty. Plus the 3rd step, overwriting the IDispatch VTable to allow dot access to members of the type library (I really like that tbh, letting the compiler/interpreter create the DISPARAMS structure for you) would be much easier if VBA classes could modify their own implementation of IDispatch directly.

**Comment by mwolfe02** [comment]
When I was put on the spot, I suggested using the [`CallByName` function](https://docs.microsoft.com/en-us/office/vba/Language/Reference/user-interface-help/callbyname-function) to accomplish this, with the caveat that you would have no compile-time type safety or error checking.  Would that work or did I completely underestimate the complexity of the task?  (Based on this thread, I suspect it's the latter 😂)

**Comment by bclothier** [comment]
The problem with the `CallByName` is that to do this, you need the object in the first place. If the object is from the host's application object model, no problem. But if it's a VBA class, then that is a problem. In past, Rubberduck sort of worked around this by using `Application.Run()` but unfortunately, Microsoft being Microsoft, it does not work the same way across different VBA hosts. For example, Access only allows you to invoke globally visible module procedures, effectively forcing you to have a list of unique names. Excel, OTOH, allowed you to disambiguate ambiguous names by fully qualifying the procedure (e.g. `VBAProject1.Module1.Foo` vs. `VBAProject2.Module1.Foo` or `VBAProject1.Module2.Foo`. But from a standpoint of authoring an unit testing framework, that was a horrid workaround and does not even address the need to test objects. 

That said, I admit I was confused by the question because normally, add-ins are meant to be used as a developer's tool, rather than to be used at runtime, so it's not typically expected to be able to interact with the user's code at the runtime. Unit testing framework is obviously an exception.

**Comment by mwolfe02** [comment]
Off the top of my head, I couldn't remember whether CallByName was part of the Access Application object model or whether it was part of the VBA standard library.  I seemed to remember there was an Access-specific feature that wasn't available from other VBA hosts, but I couldn't think of it on the spot.  I think the [Application.Eval](https://docs.microsoft.com/en-us/office/vba/api/access.application.eval) method could be used for this purpose.  Of course, that wouldn't work in other non-Access environments, but seeing as DevCon was an Access-specific conference, that might have been the simpler and easier approach.

That said, I was with you, @bclothier, in that I wasn't sure what the questioner's potential use case might have been.  I think I mentioned in my answer that what he was asking was backward from the typical host-to-addin communication.  As you say, unit testing is an obvious exception, but I didn't get the sense that was what the questioner had in mind.

**Comment by Greedquest** [comment]
Also I don't remember, do any of the built in methods let you call private subs? Because the RD framework/that VBA I posted does.

**Comment by mburns08109** [comment]
Points to: https://github.com/WaynePhillipsEA/twinbasic/issues/781  <-- where this is exactly what I was referring to.

**Comment by Greedquest** [comment]
One of the steps here is to take the `ObjPtr(wb.VBProject.References)` and dereference it into a struct
```
Public Type VBEReferencesObj
    vTable1 As LongPtr                           'To _References vtable
    vTable2 As LongPtr
    vTable3 As LongPtr
    object1 As LongPtr
    object2 As LongPtr
    typeLib As LongPtr
    placeholder1 As LongPtr
    placeholder2 As LongPtr
    RefCount As LongPtr
End Type

Dim refData As VBEReferencesObj
CopyMemory refData, ByVal ObjPtr(VBProject.References), LenB(refData)
```

Where `typeLib` is a pointer to one of these:

```
Public Type VBETypeLibObj
    vTable1 As LongPtr                           'To ITypeLib vtable
    vTable2 As LongPtr
    vTable3 As LongPtr
    Prev As LongPtr
    '@Ignore KeywordsUsedAsMember: Looks nice, sorry ThunderFrame
    Next As LongPtr
End Type
```

Which is a node in a doubly linked list of objects.

The issue is, tB is running in a different process and uses COM marshalling to get `ObjPtr(wb.VBProject.References)`  - i.e. the pointer points to a local proxy object, not the actual `References` Object. Therefore (if I understand correctly) the dereferencing step to get a `VBEReferencesObj` and all that follows is essentially meaningless since the layout in memory of the local proxy object does not have to be the same as the remote `References` object.

How should I declare this in tB to marshall the pointers in a sensible way?

**Comment by Greedquest** [comment]
Done! https://github.com/Greedquest/vbInvoke

```vba
With CreateExtendedModuleAccessor("moduleName", ActiveVBProject)
    .Foo
End With
```
